server/protocols/webadmin/webadmin-data/src/main/java/org/apache/james/webadmin/routes/MappingRoutes.java

```diff
@@ -0,0 +1,101 @@
+ * Unless required by applicable law or agreed to in writing,   *
+ * software distributed under the License is distributed on an  *
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY       *
+ * KIND, either express or implied.  See the License for the    *
+ * specific language governing permissions and limitations      *
+ * under the License.                                           *
+ ****************************************************************/
+
+package org.apache.james.webadmin.routes;
+
+import java.util.List;
+
+import javax.inject.Inject;
+import javax.ws.rs.GET;
+import javax.ws.rs.Path;
+import javax.ws.rs.Produces;
+
+import org.apache.commons.lang3.tuple.Pair;
+import org.apache.james.rrt.api.RecipientRewriteTable;
+import org.apache.james.rrt.api.RecipientRewriteTableException;
+import org.apache.james.webadmin.Constants;
+import org.apache.james.webadmin.Routes;
+import org.apache.james.webadmin.dto.MappingValueDTO;
+import org.apache.james.webadmin.utils.ErrorResponder;
+import org.apache.james.webadmin.utils.JsonTransformer;
+import org.eclipse.jetty.http.HttpStatus;
+
+import com.github.steveash.guavate.Guavate;
+import com.google.common.collect.ImmutableListMultimap;
+
+import io.swagger.annotations.Api;
+import io.swagger.annotations.ApiOperation;
+import io.swagger.annotations.ApiResponse;
+import io.swagger.annotations.ApiResponses;
+import spark.Request;
+import spark.Response;
+import spark.Service;
+
+@Api(tags = "Mappings")
+@Path(MappingRoutes.BASE_PATH)
+@Produces(Constants.JSON_CONTENT_TYPE)
+public class MappingRoutes implements Routes {
+
+    static final String BASE_PATH = "/mappings";
+
+    private final JsonTransformer jsonTransformer;
+    private final RecipientRewriteTable recipientRewriteTable;
+
+    @Inject
+    MappingRoutes(JsonTransformer jsonTransformer, RecipientRewriteTable recipientRewriteTable) {
+        this.jsonTransformer = jsonTransformer;
+        this.recipientRewriteTable = recipientRewriteTable;
+    }
+
+    @Override
+    public String getBasePath() {
+        return BASE_PATH;
+    }
+
+    @Override
+    public void define(Service service) {
+        service.get(BASE_PATH, this::getMappings, jsonTransformer);
+    }
+
+    @GET
+    @Path(BASE_PATH)
+    @ApiOperation(value = "getting all mappings in RecipientRewriteTable")
+    @ApiResponses(value = {
+        @ApiResponse(code = HttpStatus.OK_200, message = "OK", response = List.class)
+    })
+    private ImmutableListMultimap<String, MappingValueDTO> getMappings(Request request, Response response) {
+        try {
+            return recipientRewriteTable.getAllMappings()
+                .entrySet()
+                .stream()
+                .flatMap(entry -> entry.getValue().asStream()
+                    .map(mapping -> Pair.of(entry.getKey(), mapping)))
+                .collect(Guavate.toImmutableListMultimap(
+                    pair -> pair.getLeft().asString(),
```

Comment:
You can directly do it in the `map()`

Reply:
sorry, I didn't understand your comment. Could you explain it clearer?

Reply:
Of course, currently you spread your transformation in two place:

 * `map` where you get a first `Pair<T, U>` (I do not know what they really are)
 * `collect` where you transform them again in `Pair<String, MappingValueDTO>`

I'd like a single transformation step, so you'll be able to write:

```
.collect(Guavate.toImmutableListMultimap(Pair::getLeft, Pair::getRight))
```

Reply:
thank you, I got it